### ISSUE TITLE
addChannelPermission Function Does Not Support Large Icon IDs

### ISSUE THREAD

**Issue by KaiAnti** [TheHolyWaffle/TeamSpeak-3-Java-API#425]
Problem Description:
The addChannelPermission function currently accepts the permission value as an int. However, when setting the i_icon_id permission for channels in TeamSpeak, some icon IDs exceed the maximum value of a 32-bit integer (Integer.MAX_VALUE). This causes the function to fail for large icon IDs.

Since TeamSpeak assigns icon IDs automatically and they cannot be modified by the user, this limitation prevents the use of certain icons with large IDs. This is a significant restriction, especially for larger servers or servers with a high number of uploaded icons.

Proposed Solution:
To address this issue, the parameter type for the permission value in addChannelPermission should be changed from int to long. This would allow the function to handle the full range of TeamSpeak's icon IDs and resolve compatibility issues.

Steps to Reproduce:
1. Attempt to set a channel's i_icon_id using addChannelPermission with an icon ID larger than Integer.MAX_VALUE.
2.Observe that the function fails due to the value exceeding the int limit.

Expected Behavior:
The function should successfully assign i_icon_id permissions for all valid TeamSpeak icon IDs, regardless of their size.

Environment:
Java Library: 1.3.1
TeamSpeak Server Version: 3.13.7

Additional Notes:
This change would improve compatibility and user experience for managing TeamSpeak servers programmatically. Given that long can handle larger numbers, it aligns better with TeamSpeak's ID assignment system.

the command:
Load.api.addChannelPermission(newChannelID, "i_icon_id", 2280211472);
		
the Error:
2024-12-13 20:43:36.583 [ERROR] Event listener threw an exception
java.lang.NumberFormatException: For input string: "2280211472"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:588)
	at java.base/java.lang.Integer.parseInt(Integer.java:685)
	at Java_TS3_Bot/nofame_esports_Bot_events.selection_channel.create_selection_tempchannel(selection_channel.java:84)
	at Java_TS3_Bot/nofame_esports_Bot_events.Event$1.onClientMoved(Event.java:156)
	at com.github.theholywaffle.teamspeak3@1.3.1-with-dependencies/com.github.theholywaffle.teamspeak3.api.event.ClientMovedEvent.fire(ClientMovedEvent.java:55)
	at com.github.theholywaffle.teamspeak3@1.3.1-with-dependencies/com.github.theholywaffle.teamspeak3.EventManager$ListenerTask.run(EventManager.java:150)
	at com.github.theholywaffle.teamspeak3@1.3.1-with-dependencies/com.github.theholywaffle.teamspeak3.TS3Query.lambda$submitUserTask$0(TS3Query.java:282)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1575)

**Comment by astrolamb-gaming** [comment]
@KaiAnti While I don't disagree that it would be beneficial to change this, since the getter returns a long, I might have a workaround for you.
I was able to successfully use the following:
```Load.api.addChannelPermission(newChannelID, "i_icon_id", (int) 2280211472L);```

**Comment by rogermb** [comment]
Permission values / icon IDs are 32-bit integers, both in TeamSpeak3 and in this API. The only difference is that these IDs are displayed as unsigned (0 .. 2<sup>32</sup>-1) in your TS3 client and the server query interface, and as signed (-2<sup>31</sup> .. 2<sup>31</sup>-1) in Java, because Java does not have any unsigned integer types.

To print this 32-bit integer as the unsigned value you expect, use [Integer.toUnsignedString](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#toUnsignedString-int-). To parse an unsigned 32-bit integer from a string, use [Integer.parseUnsignedInt](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#parseUnsignedInt-java.lang.String-). If you just have a constant value in your code somewhere, @astrolamb-gaming's suggested "use long & cast to int" method is a great option.

---

I agree that it would've likely been a cleaner solution to just use longs everywhere to work around Java's lack of unsigned numeric types from the start. And it's especially ugly that we have icon ID getters that return longs, but then only have permission setters that accept ints.

Perhaps we should add some dedicated `set<X>IconId` methods to TS3Api / TS3ApiAsync that down-cast to int and end up calling the respective `add<X>Permission` methods. But then that's like a dozen more methods added to those classes, just to save on having to downcast to int one or two times in client applications.

¯\\\_(ツ)_/¯

**Comment by astrolamb-gaming** [comment]
Yeah, I was poking around to see what it would take to implement longs for icons. KeyValueParam has constructors that use strings, ints, longs, and booleans, but the higher level api uses ints in this case, so having the long conversion doesn't help us.
I've been wondering if, since it all gets converted to strings anyways, maybe the permission value should always just be a string, forcing the user to convert to string when using it.

**Comment by KaiAnti** [comment]
@astrolamb-gaming Thank you for pointing out this workaround! While your solution works for values within the int range (e.g., 2280211472L), I’ve encountered an issue with numbers that exceed this range, such as 3624801561L. These values cannot simply be cast to an int because they fall outside the valid range of -2,147,483,648 to 2,147,483,647.

In our case, this limitation affects 10 channels whose icons the bot cannot handle, while it works perfectly for 3 channels. The root cause of this issue is the restriction to int, which could be avoided if the API allowed numbers to be passed as long directly, without requiring cumbersome conversions or workarounds.

Unfortunately, I don’t have the technical expertise to implement such a change in the API myself. It would be amazing if someone with the necessary background could look into this. Extending the API to support long values would not only make it more flexible but also simpler and more robust for scenarios like this.

Thank you in advance for your support!

**Comment by rogermb** [comment]
@KaiAnti Sorry for taking a while to get back to you, work started again after my Christmas break, so I was busy.

> I’ve encountered an issue with numbers that exceed this range, such as 3624801561L. These values cannot simply be cast to an int because they fall outside the valid range of -2,147,483,648 to 2,147,483,647.

Huh? Again, TS3 also only uses 32-bit ints. You can just cast it and get the equivalent signed int. It's fine. TS3 handles it.

For example, I have this "muted" icon that exceeds 2<sup>32</sup>-1:
![image](https://github.com/user-attachments/assets/9dd14d81-4a10-45ca-aa6e-ba4fbd117a62)

Here I created a new channel, Test, with channel ID 1944:
![image](https://github.com/user-attachments/assets/c72375ac-20aa-4500-bc44-2d4a485fe2c4)

Now let's set this icon ID using the API:
```java
api.addChannelPermission(1944, "i_icon_id", (int) 3815707113L);
```

This results in this command being sent:
```
2025-01-12 14:46:20.592 [DEBUG] [channeladdperm] > channeladdperm cid=1944 permsid=i_icon_id permvalue=-479260183
2025-01-12 14:46:20.594 [DEBUG] [channeladdperm] < error id=0 msg=ok
```

And if we check the TS3 server, the icon has been set:
![image](https://github.com/user-attachments/assets/3b6f85aa-cb29-4a5d-874b-f2a061aa5fc1)

And if we get the permission value using the TS3 API again, we get:
```
2025-01-12 14:46:20.596 [DEBUG] [channelpermlist] < cid=1944 permsid=i_icon_id permvalue=-479260183 permnegated=0 permskip=0
```

Basically, things seem to work perfectly on my end, and I honestly don't quite understand where the confusion is coming from. Are you using some other TS3 server implementation that handles these commands / icon IDs differently?

**Comment by rogermb** [comment]
And to add to that, even if we'd go out of our way to send the unsigned integer value 3815707113 to the server, it wouldn't matter, TS3 still interprets permission values as signed 32-bit integers:

```
2025-01-12 14:55:03.612 [DEBUG] [channeladdperm] > channeladdperm cid=1944 permsid=i_icon_id permvalue=3815707113
2025-01-12 14:55:03.612 [DEBUG] [channeladdperm] < error id=0 msg=ok
2025-01-12 14:55:03.613 [DEBUG] [channelpermlist] > channelpermlist cid=1944 -permsid
2025-01-12 14:55:03.614 [DEBUG] [channelpermlist] < cid=1944 permsid=i_icon_id permvalue=-479260183 permnegated=0 permskip=0
```
